File: 49-group-anagrams/49-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> m; 
        vector<vector<string>> ans; 
        
        for(int i=0;i<strs.size();i++){
                string p = strs[i]; 
                sort(p.begin(), p.end()); 
                m[p].push_back(strs[i]);
        }
        
        for(auto it= m.begin();it!=m.end();it++){
            ans.push_back(it->second);
        }
     
        return ans;
    }
};

File: 49-group-anagrams/49-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> result = new ArrayList<List<String>>();
        
        Map<String, List<String>> groups = new HashMap<String, List<String>>();
        for (int i = 0; i < strs.length; i++){
            String str = strs[i];
            String key = buildAnagramsKey(str);
            
            if (!groups.containsKey(key)){
                groups.put(key, new ArrayList<String>());
            }
            groups.get(key).add(str);
        }
        
        for (Map.Entry<String, List<String>> pair : groups.entrySet()){
            result.add(pair.getValue());
        }
        
        return result;
    }
    
    public String buildAnagramsKey(String str){
        int[] map = new int[26];
        
        for (Character c : str.toCharArray()){
            map[c - 'a'] += 1;
        }
        StringBuilder build = new StringBuilder();
        for (int i = 0; i < 26; i++){
            build.append(map[i]);
            build.append((char) (i + 'a'));
        }
        return build.toString();
    }
}